sap/note/parameter.go

```diff
@@ -0,0 +1,213 @@
+		if err != nil {
+			log.Printf("Failed to store start values for parameter file '%s' for parameter '%s'", GetPathToParameter(param), param)
+		}
+	}
+}
+
+// Add note parameter values to the state file.
+func AddParameterNoteValues(param, value, noteID string) {
+	pEntries := ParameterNotes {
+		AllNotes: make([]ParameterNoteEntry, 0, 64),
+	}
+	pEntries = GetSavedParameterNotes(param)
+	if len(pEntries.AllNotes) != 0 && !NoteInParameterList(noteID, pEntries.AllNotes) {
+		log.Printf("Write note '%s' parameter value '%s' to file '%s'", noteID, value, GetPathToParameter(param))
+		// file exis
+		pEntry := ParameterNoteEntry{
+			NoteId:   noteID,
+			Value:    value,
+		}
+		pEntries.AllNotes = append(pEntries.AllNotes, pEntry)
+		err := StoreParameter(param, pEntries, true)
+		if err != nil {
+			log.Printf("Failed to store note '%s' values for parameter file '%s' for parameter '%s'", noteID, GetPathToParameter(param), param)
+		}
+	}
+}
+
+// Read content of stored paramter states. Return the content as ParameterNotes
+func GetSavedParameterNotes(param string) (ParameterNotes) {
+	pEntries := ParameterNotes {
+		AllNotes: make([]ParameterNoteEntry, 0, 64),
+	}
+	content, err := ioutil.ReadFile(GetPathToParameter(param))
+	if err != nil {
+		return pEntries
+	}
+	err = json.Unmarshal(content, &pEntries)
+	return pEntries
+}
+
+// read all saved parameters from the state directory
+// fill structure app.AllParameters
+func GetAllSavedParameters() (map[string]ParameterNotes) {
+	params := make(map[string]ParameterNotes)
+	allParams, err := ListParams()
+	if err == nil {
+		for _, param := range allParams {
+			pEntries := GetSavedParameterNotes(param)
+			if len(pEntries.AllNotes) == 0 {
+				return params
+			}
+			params[param] = pEntries
+		}
+	}
+	return params
+}
+
+// Store parameter values to state directory
+// Write a json file with the name of the given parameter containing the 
+// applied noteIDs for this parameter and the associated parameter values
+func StoreParameter(param string, obj ParameterNotes, overwriteExisting bool) error {
+	content, err := json.Marshal(obj)
+	if err != nil {
+		return err
+	}
+	if err = os.MkdirAll(SaptuneParameterStateDir, 0755); err != nil {
+		return err
+	}
+	if _, err := os.Stat(GetPathToParameter(param)); os.IsNotExist(err) || overwriteExisting {
+		return ioutil.WriteFile(GetPathToParameter(param), content, 0644)
+	}
+	return nil
+}
+
+// for a given noteID get the position in the slice AllNotes
+// return the position of the note within the slice.
+// do not sort the slice
+func PositionInParameterList(noteID string, list []ParameterNoteEntry) int {
+	for cnt, note := range list {
+		if note.NoteId == noteID {
+			return cnt
+		}
+	}
+	return 0
+}
+
+// Revert parameter value and remove noteID reference from parameter file
+// return value of parameter, if needed to change, empty string else
+func RevertParameter(param string, noteID string) (string) {
```

Comment:
@angelabriel, especially the revert part is getting more complex now. Would it deserve a better unit testing coverage, or is it already covered on the changes you did on the `app_test.go`?

Reply:
your right. At the moment the revert part is getting more complex. But it's now aligned with our new sap note definition handling (different note definition files can set the same system parameter but with different values) and from this point of view, it's now more clear. 
The changes for app_test.go cover only some test cases. I need to add some more tests in parameter_test.go, which will be done in the current sprint.

Reply:
Ok, crystal clear to me.